### PR TITLE
feat(relay): reloading certificate resolver

### PR DESCRIPTION
**Pull request by Arqu**
## Description

This sets us up so we can have manually/externally managed certs that will reload on a daily basis. This should close https://github.com/n0-computer/iroh/issues/1108 and we should follow up on the ops side to utilize this where needed.

Ok, this turned out to be much grosser than expected. 
- I don't like the new dependencies we introduce
- I don't like the loader setup
- I'm unsure whether to keep the reloading resolver in the `iroh` repo or move it into `tokio-rustls-acme` (think it should move over, but for the sake of this discussion, included it here)
- Should I do anything with the `_handle` for the reloader in regards to shutting down?

Most of this still feels necessary. I could maybe feature flag it?
I had a really hard time trying to extract the loader creation into a standalone func which also aligns all the traits and error types so the thing compiles.

Suggestions are welcome :)

Example config:
```
enable_relay = true
http_bind_addr = "[::]:80"
enable_stun = true
stun_bind_addr = "[::]:3478"
enable_metrics = true
metrics_addr = "127.0.0.1:9090"

[tls]
manual_cert_path="certificate.der"
manual_key_path="private_key.der"
cert_mode = "Reloading"
```

## Change checklist

- [ ] Self-review.
- [ ] Documentation updates following the [style guide](https://rust-lang.github.io/rfcs/1574-more-api-documentation-conventions.html#appendix-a-full-conventions-text), if relevant.
- [ ] Tests if relevant.
- [ ] All breaking changes documented.

### CODE REVIEW

**Comment by Arqu** [comment]
From @flub in discord
> i have some comments on the config structure, code (and bikeshedding 😉 ), but the principle of it doesn't seem that gross?

> IIUC this reads certs from a file at given intervals.  which seems fine
Correct
> i'm also not sure why you think this might belong in tokio-acme-rustls, it doesn't seem to have anything acme about it?
I guess what I wanted it to be is mostly it's own separate thing. It's both generally useful outside of iroh and also makes the relay codebase less cluttered. I also expect this to crop up with other services we're running, I'd rather not copy it over.
> wrt the comments, my biggest structural one is probably why this is separate from Manual as it seems at some level this is > just manual + an interval?
I agree, the `CertMode` handling should be separate, but bellow that it's the same as `Manual` anyways. 

Just moving the discussion over here to make things a bit easier.

Thanks for the input!

**Comment by flub** [comment]
> > i'm also not sure why you think this might belong in tokio-acme-rustls, it doesn't seem to have anything acme about it?
> 
> I guess what I wanted it to be is mostly it's own separate thing. It's both generally useful outside of iroh and also makes the relay codebase less cluttered. I also expect this to crop up with other services we're running, I'd rather not copy it over.

This is not that much code and I think it fits just fine inside the relay server for now.  Once we have to start copy-pasting it we can re-visit that I guess.

File: iroh-relay/src/server/resolver.rs [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+    /// Perform the initial load and construct the [`ReloadingResolver`].
+    pub async fn init(loader: Loader, interval: Duration) -> Result<Self> {
+        let (reloadable, _) = Reloadable::init_load(loader)
+            .await
+            .map_err(|_| anyhow!("Failed to load the certificate"))?;
+        let reloadable = Arc::new(reloadable);
+
+        // Spwan a task to reload the certificate every interval.
+        let _reloadable = reloadable.clone();
+        let _handle = tokio::spawn(async move {
```
Comment (dignifiedquire) [inline iroh-relay/src/server/resolver.rs:33]: should have a way to cancel/shutdown?

**Comment by Arqu** [comment]
Looking to get back to this tonight

File: iroh-relay/src/server/resolver.rs [hunk 2]
```diff
@@ -0,0 +1,90 @@
...
+
+/// A Certificate resolver that reloads the certificate every interval
+#[derive(Debug)]
+pub struct ReloadingResolver<Loader: Send + 'static> {
+    /// The inner reloadable value.
+    reloadable: Arc<Reloadable<CertifiedKey, Loader>>,
+    /// The handle to the task that reloads the certificate.
+    _handle: AbortOnDropHandle<()>,
+    /// Shutdown signal sender
+    _shutdown_tx: tokio::sync::oneshot::Sender<()>,
```
Comment (Arqu) [inline iroh-relay/src/server/resolver.rs:19]: Should I bubble this up into the relay so we call it on relay.shutdown() or is the AbortOnDropHandle enough for this usecase?

**Comment by Arqu** [comment]
Do we have to bump the MSRV because redb got bumped to 2.3?

**Comment by dignifiedquire** [comment]
> Do we have to bump the MSRV because redb got bumped to 2.3?

what's their new MSRV? we probably want to bump it to normalize cargo.lock to v4

**Comment by Arqu** [comment]
> > Do we have to bump the MSRV because redb got bumped to 2.3?
> 
> what's their new MSRV? we probably want to bump it to normalize cargo.lock to v4

[1.81](https://github.com/cberner/redb/commit/bb4415413906bb01341c7fb852d2b0af17046d9b)

**Comment by dignifiedquire** [comment]
lets bump MSRV to 1.81 then, this will make live easier for us

**Comment by dignifiedquire** [comment]
but in a new PR

**Comment by dignifiedquire** [comment]
https://github.com/n0-computer/iroh/issues/3031

**Comment by Arqu** [comment]
Rest of the PR good now?

File: iroh-relay/src/server/resolver.rs [hunk 3]
```diff
@@ -0,0 +1,90 @@
...
+            let mut interval = tokio::time::interval(interval);
+            tokio::select! {
+                _ = async {
+                    loop {
+                        interval.tick().await;
+                        let _ = _reloadable.reload().await;
+                        tracing::info!("Reloaded the certificate");
+                    }
+                } => {},
+                _ = shutdown_rx => {
```
Comment (dignifiedquire) [inline iroh-relay/src/server/resolver.rs:48]: we usually use cancellation token for this

**Comment by dignifiedquire** [comment]
MSRV fix is in now

**Review by dignifiedquire — APPROVED** [review]

File: iroh-relay/src/server/resolver.rs [hunk 4]
```diff
@@ -0,0 +1,93 @@
...
+use anyhow::{anyhow, Result};
+use reloadable_state::Reloadable;
+use rustls::{
+    server::{ClientHello, ResolvesServerCert},
+    sign::CertifiedKey,
+};
+use tokio_util::{sync::CancellationToken, task::AbortOnDropHandle};
+
+/// The default certificate reload interval.
+pub const DEFAULT_CERT_RELOAD_INTERVAL: u64 = 60 * 60 * 24;
```
Comment (flub) [inline iroh-relay/src/server/resolver.rs:12]: No units!  Maybe make it a `Duration`?  Or is that still not `const`?  Should definitely document the units though.

**Comment by flub** [comment]
lol, i miss the review window again :smile:

**Comment by Arqu** [comment]
I'll do a follow up, just for you :D

**Pull request merged**